### PR TITLE
[Validator] Add "is_valid" function to the Expression constraint

### LINKED ISSUE #11940: [Validator] Validate from within Expression
The issue #11586 could be solved by providing a `valid()` method in the Expression constraint (as alias for `$context->getValidator()->validate()`, but returning a boolean):

``` php
/**
 * @Expression("valid(this.author) or valid(this.creator)", message="Author or creator must be valid!")
 */
```

Additionally, a "constraints" option could be added which allows to pass constraints that can be used within the expression:

``` php
/**
 * @Expression(
 *     "valid(value, regex) or valid(value, threeOrMore)",
 *     message="Value must match regex or be greater than 3!",
 *     constraints={
 *         "regex" = @Regex("/here-is-my-regex/"),
 *         "threeOrMore" = @GreaterThan(3),
 *     }
 * )
 */
```

Opinions?

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix https://github.com/symfony/symfony/issues/11940
| License       | MIT
| Doc PR        | TODO

The main benefit of this feature is to be able to support "or" constraints but it generally open the usage of this constraint to many cool cases.

`is_valid` accepts an unlimited number of arguments of 3 kinds:

1. Constraints (resolved in any way)
```php
/**
 * @Expression("is_valid(a) or is_valid(b)", values={"a": @Type('string'), "b": @Type('int')})
 */
```

```php
/**
 * @Expression("is_valid(this.method(), a)", values={"a": @Type('string')})
 */

// ...
public function method() { return new IdenticalToo('foo'); }
```

2. Array of constraints (resolved in any way)
```php
/**
 * @Expression("is_valid(a) or ...", values={"a": {@Type('int'), @GreaterThan(10)}})
 */
```

3. Strings that represents properties when validating an object
```php
/**
 * @Expression("not is_valid("foo") or not is_valid("bar")})
 */

// ...
public $foo;
public $bar;
```

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/ExpressionValidator.php [hunk 1]
```diff
@@ -65,13 +68,87 @@ public function validate($value, Constraint $constraint)
 
     private function getExpressionLanguage()
     {
-        if (null === $this->expressionLanguage) {
+        if (!$this->expressionLanguage instanceof ExpressionLanguage) {
```
Comment (fancyweb) [inline src/Symfony/Component/Validator/Constraints/ExpressionValidator.php]: This is a bug fix on 4.4, `$this->expressionLanguage` can be anything. I will extract it in a separate PR if this one doesn't end up being merged.

File: src/Symfony/Component/Validator/Constraints/ExpressionValidator.php [hunk 2]
```diff
@@ -65,13 +68,87 @@ public function validate($value, Constraint $constraint)
...
+
+                    $properties[] = $argument;
+
+                    continue;
+                }
+
+                throw new ConstraintDefinitionException(sprintf('The "is_valid" function only accepts instances of "%s", arrays of "%s", or strings that represent properties paths (when validating an object), "%s" given.', Constraint::class, Constraint::class, \is_object($argument) ? \get_class($argument) : \gettype($argument)));
+            }
+
+            if (!$constraints && !$properties) {
```
Comment (fancyweb) [inline src/Symfony/Component/Validator/Constraints/ExpressionValidator.php]: This check is useless, I will remove it.

File: src/Symfony/Component/Validator/Constraints/ExpressionValidator.php (hunk 1 above)
Comment (yceruto) [reply src/Symfony/Component/Validator/Constraints/ExpressionValidator.php]: Fixed in https://github.com/symfony/symfony/pull/33831

File: src/Symfony/Component/Validator/Constraints/ExpressionValidator.php [hunk 3]
```diff
@@ -65,13 +68,87 @@ public function validate($value, Constraint $constraint)
...
         return $this->expressionLanguage;
     }
+
+    private function addIsValidFunction(): void
+    {
+        if ($this->expressionLanguage->hasFunction('is_valid')) {
+            return;
+        }
+
+        $this->expressionLanguage->register('is_valid', function () {
```
Comment (yceruto) [inline src/Symfony/Component/Validator/Constraints/ExpressionValidator.php]: This code must be moved to a dedicated class `ExpressionLanguageProvider` by implementing `ExpressionFunctionProviderInterface` and later `$this->expressionLanguage->registerProvider(...)`.

File: src/Symfony/Component/Validator/Tests/Fixtures/FakeMetadataFactory.php [hunk 4]
```diff
@@ -62,6 +62,8 @@ public function hasMetadataFor($class): bool
     public function addMetadata($metadata)
     {
         $this->metadatas[$metadata->getClassName()] = $metadata;
+
+        return $this;
```
Comment (yceruto) [inline src/Symfony/Component/Validator/Tests/Fixtures/FakeMetadataFactory.php]: Unrelated change? if not, you should add the phpdoc block too: `@return $this`.

**Review by yceruto — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Validator/Constraints/ExpressionValidator.php (hunk 1 above)
Comment (fancyweb) [reply src/Symfony/Component/Validator/Constraints/ExpressionValidator.php]: Thanks, reverting here then.

File: src/Symfony/Component/Validator/Tests/Fixtures/FakeMetadataFactory.php (hunk 4 above)
Comment (fancyweb) [reply src/Symfony/Component/Validator/Tests/Fixtures/FakeMetadataFactory.php]: Used it for a test but not the case anymore, I just revert this change.

**Comment by jvasseur** [comment]
The API feals really strange, I would expect calling `is_valid(a)` will test if `a` is valid instead of expecting `a` to be a constraint.

We could expose an API similar to the `ValidatorInterface::validate` function: `is_valid(value, constraint(s) =  null)` 

With this API your use cases would become:

```php
/**
 * Constraints:
 * @Expression("is_valid(value, a) or is_valid(value, b)", values={"a": @Type('string'), "b": @Type('int')})
 *
 * Properties:
 * @Expression("not is_valid(value.foo) or not is_valid(value.bar)")
 */
```

It's a bit more verbose but I think it's a lot more clear what is validated and against which constraints.

**Comment by fancyweb** [comment]
> `is_valid(value, a)`

I agree it's more understandable but the first argument will always be `value` then 😕 And it opens  the ability to validate other values of objects (eg: `is_valid(this.getFoo(), a)`. I guess it's not the goal of such a feature.

> `is_valid(value.foo)`

Not possible because the function would get the resolved value, not the property. The passed arg has to be a string. `this.foo` could be an alternative.

**Comment by jvasseur** [comment]
> And it opens the ability to validate other values of objects (eg: is_valid(this.getFoo(), a). I guess it's not the goal of such a feature.

I don't see why it shouldn't be allowed.

> Not possible because the function would get the resolved value

That's what I wanted to do, the function would get the resolved value and validated it. But I guess what you wanted is validate it against the constraints defined in the object (the same way `validateProperty` works).

Maybe we could allow a `is_valid(object, propertyName)` form to allow validating properties.

**Comment by fancyweb** [comment]
@jvasseur After some more reflexion I think you are right: `is_valid(value, ...constraints)` looks better for DX, especially since `value` is already a value already available in the expression. And I think we can actually support both ways thanks to variadic arguments by checking if the first one is an instance of constraint. I drop support for array of constraints -> it can be replaced by the `All` constraint.

> That's what I wanted to do, the function would get the resolved value and validated it
For objects we can but for anything else, there would be no constraints to check against, isn't it?

>  I guess what you wanted is validate it against the constraints defined in the object (the same way validateProperty works).
Yes! 

> Maybe we could allow a is_valid(object, propertyName) form to allow validating properties.
Or we can just have another function `is_property_valid(object, ...property)` with object being optional by checking the first argument again.

**Comment by fabpot** [comment]
@fancyweb I see that you changed your mind about how to solve the problem. So you want to update this PR accordingly?

**Comment by fancyweb** [comment]
I need to refocus on this work as I left it "on the side". I'm closing and I will maybe reopen a new PR in the future.
